src/components/OrderBookWidget.tsx

```diff
@@ -84,67 +67,161 @@ enum Offer {
+      amount
+        .times(TEN_BIG_NUMBER.pow(new BigNumber(DEFAULT_PRECISION)))
+        // we don't want any decimals
+        // before passing into BN conversion
+        .decimalPlaces(0)
+        .toString(10),
+    ),
+    precision: DEFAULT_PRECISION,
+    decimals: 6,
+    smallLimit: '0',
+  })
 }
 
 /**
  * This method turns the raw data that the backend returns into data that can be displayed by the chart.
  * This involves aggregating the total volume and accounting for decimals
  */
 const processData = (
-  list: RawItem[],
+  rawPricePoints: RawPricePoint[],
   baseToken: TokenDetails,
   quoteToken: TokenDetails,
   type: Offer,
-): ProcessedItem[] => {
-  let totalVolume = 0
+): PricePointDetails[] => {
   const isBid = type == Offer.Bid
-  return (
-    list
-      // Account fo decimals
-      .map(element => {
-        return {
-          price: element.price / 10 ** (quoteToken.decimals - baseToken.decimals),
-          volume: element.volume / 10 ** baseToken.decimals,
-        }
-      })
-      // Filter tiny orders
-      .filter(e => e.volume > 0.01)
-      // Accumulate totalVolume
-      .map(e => {
-        const previousTotalVolume = totalVolume
-        totalVolume += e.volume
-        return {
-          price: e.price,
-          volume: e.volume,
-          totalVolume,
-          // Amcharts draws step lines so that the x value is centered (Default). To correctly display the orderbook, we want
-          // the x value to be at the left side of the step for asks and at the right side of the step for bids.
-          //
-          //    Default            Bids          Asks
-          //            |      |                        |
-          //   ---------       ---------       ---------
-          //  |                         |      |
-          //       x                    x      x
-          //
-          // For asks, we can offset the "startLocation" by 0.5. However, Amcharts does not support a "startLocation" of -0.5.
-          // For bids, we therefore offset the curve by -1 (expose the previous total volume) and use an offset of 0.5.
-          // Otherwise our steps would be off by one.
-          askValueY: isBid ? null : totalVolume,
-          bidValueY: isBid ? previousTotalVolume : null,
-        }
-      })
+  const quoteTokenDecimals = quoteToken.decimals
+  const baseTokenDecimals = baseToken.decimals
+
+  // Convert RawPricePoint into PricePoint:
+  //  Raw items use number (floats) and are given in "atoms"
+  //  Normalized items use decimals (BigNumber) and are given in natural units
+  let pricePoints: PricePoint[] = rawPricePoints.map(pricePoint =>
+    _toPricePoint(pricePoint, quoteTokenDecimals, baseTokenDecimals),
+  )
+
+  // Filter tiny orders
+  pricePoints = pricePoints.filter(pricePoint => pricePoint.volume.gt(SMALL_VOLUME_THRESHOLD))
+
+  // Convert the price points that can be represented in the graph (PricePointDetails)
+  const { points } = pricePoints.reduce(
```

Comment:
While this is all nicely declarative, I feel like orderbook is sometimes slow and these 3 loops can be done in one inside reduce.
Just a thought. Feel free to ignore

Reply:
https://github.com/gnosis/dex-react/pull/1069